File: lib/utils/pinyin_tone.dart
import 'package:flutter/material.dart';

/// Colors pinyin text by tone: 1st=blue, 2nd=green, 3rd=yellow, 4th=red,
/// neutral=grey. Works by splitting each run of pinyin letters into
/// syllables (using a standard initials/finals table) and then reading the
/// tone off whichever vowel in that syllable carries a diacritic.
class PinyinTone {
  PinyinTone._();

  static const Map<int, Color> colors = {
    1: Color(0xFF1E88E5), // blue
    2: Color(0xFF43A047), // green
    3: Color(0xFFF9A825), // yellow (darkened for legibility)
    4: Color(0xFFE53935), // red
    5: Color(0xFF9E9E9E), // grey - neutral tone
  };

  static const Map<String, int> _toneOfCharLower = {
    'ā': 1, 'á': 2, 'ǎ': 3, 'à': 4,
    'ē': 1, 'é': 2, 'ě': 3, 'è': 4,
    'ī': 1, 'í': 2, 'ǐ': 3, 'ì': 4,
    'ō': 1, 'ó': 2, 'ǒ': 3, 'ò': 4,
    'ū': 1, 'ú': 2, 'ǔ': 3, 'ù': 4,
    'ǖ': 1, 'ǘ': 2, 'ǚ': 3, 'ǜ': 4,
  };

  static const Map<String, String> _plainOfCharLower = {
    'ā': 'a', 'á': 'a', 'ǎ': 'a', 'à': 'a',
    'ē': 'e', 'é': 'e', 'ě': 'e', 'è': 'e',
    'ī': 'i', 'í': 'i', 'ǐ': 'i', 'ì': 'i',
    'ō': 'o', 'ó': 'o', 'ǒ': 'o', 'ò': 'o',
    'ū': 'u', 'ú': 'u', 'ǔ': 'u', 'ù': 'u',
    'ǖ': 'v', 'ǘ': 'v', 'ǚ': 'v', 'ǜ': 'v',
    'ü': 'v', 'Ü': 'v',
  };

  // Sentence-initial pinyin syllables are often capitalized (e.g. "Àiqíng
  // hěn zhòngyào."), so every toned vowel needs an uppercase counterpart too.
  static final Map<String, int> _toneOfChar = {
    ..._toneOfCharLower,
    for (final entry in _toneOfCharLower.entries) entry.key.toUpperCase(): entry.value,
  };

  static final Map<String, String> _plainOfChar = {
    ..._plainOfCharLower,
    for (final entry in _plainOfCharLower.entries) entry.key.toUpperCase(): entry.value,
  };

  /// Replaces toned vowels with their plain ASCII form (ǚ→v, é→e), leaving
  /// other characters untouched. Note ü/ǚ map to "v".
  static String stripTones(String s) =>
      s.split('').map((c) => _plainOfChar[c] ?? c).join();

  static final List<String> _initials = (<String>[
    'zh', 'ch', 'sh',
    'b', 'p', 'm', 'f', 'd', 't', 'n', 'l',
    'g', 'k', 'h', 'j', 'q', 'x', 'r', 'z', 'c', 's', 'y', 'w',
  ])..sort((a, b) => b.length.compareTo(a.length));

  static final List<String> _finals = (<String>[
    'iang', 'iong', 'uang', 'ueng',
    'ian', 'iao', 'ing', 'ong', 'uai', 'uan', 'van',
    'ai', 'ei', 'ao', 'ou', 'an', 'en', 'ang', 'eng', 'er', 'ue',
    'ia', 'ie', 'iu', 'in', 'ua', 'uo', 'ui', 'un', 've',
    'a', 'o', 'e', 'i', 'u', 'v',
  ])..sort((a, b) => b.length.compareTo(a.length));

  static bool _isPinyinLetter(String c) {
    if (c.isEmpty) return false;
    final code = c.codeUnitAt(0);
    final isAscii = (code >= 65 && code <= 90) || (code >= 97 && code <= 122);
    return isAscii || _toneOfChar.containsKey(c) || _plainOfChar.containsKey(c);
  }

  static String _plain(String c) {
    if (_plainOfChar.containsKey(c)) return _plainOfChar[c]!;
    return c.toLowerCase();
  }

  /// Splits one contiguous run of pinyin letters (already known to contain
  /// no spaces/punctuation) into (syllableText, toneNumber) pairs.
  static List<(String, int)> _splitRun(String run) {
    final plain = run.split('').map(_plain).join();
    final result = <(String, int)>[];
    var pos = 0;

    while (pos < run.length) {
      String matchedInitial = '';
      for (final init in _initials) {
        if (plain.startsWith(init, pos)) {
          matchedInitial = init;
          break;
        }
      }
      final afterInitial = pos + matchedInitial.length;

      String matchedFinal = '';
      for (final fin in _finals) {
        if (plain.startsWith(fin, afterInitial)) {
          matchedFinal = fin;
          break;
        }
      }

      if (matchedFinal.isEmpty) {
        final chunk = run.substring(pos, pos + 1);
        // A lone trailing "r" after a full syllable is the 儿化 (erhua)
        // suffix (e.g. "nǎr", "wánr") — it's part of the previous
        // syllable's sound, not a syllable of its own, so keep it attached
        // rather than coloring it separately as a fake neutral-tone blip.
        if ((chunk == 'r' || chunk == 'R') && result.isNotEmpty) {
          final last = result.removeLast();
          result.add(('${last.$1}$chunk', last.$2));
        } else {
          // Unexpected input we couldn't segment — fall back to a single
          // character so we always make forward progress.
          result.add((chunk, _toneOf(chunk)));
        }
        pos += 1;
        continue;
      }

      final end = afterInitial + matchedFinal.length;
      final syllable = run.substring(pos, end);
      result.add((syllable, _toneOf(syllable)));
      pos = end;
    }

    return result;
  }

  static int _toneOf(String syllable) {
    for (final c in syllable.split('')) {
      final tone = _toneOfChar[c];
      if (tone != null) return tone;
    }
    return 5; // neutral
  }

  /// Flat, in-order list of every syllable in [pinyin] as
  /// (syllableText, tone) pairs, skipping spaces/punctuation. Used to
  /// tone-color hanzi: the Nth CJK character of the paired Chinese text
  /// corresponds to the Nth syllable here (with 儿化 handled by the caller,
  /// since an erhua syllable like "wánr" covers two characters: 玩儿).
  static List<(String, int)> syllables(String pinyin) {
    final result = <(String, int)>[];
    var i = 0;
    while (i < pinyin.length) {
      if (_isPinyinLetter(pinyin[i])) {
        var j = i + 1;
        while (j < pinyin.length && _isPinyinLetter(pinyin[j])) {
          j++;
        }
        result.addAll(_splitRun(pinyin.substring(i, j)));
        i = j;
      } else {
        i++;
      }
    }
    return result;
  }

  /// Builds a list of [TextSpan]s for [pinyin], coloring each syllable by
  /// its tone while leaving spaces/punctuation in the default color.
  static List<InlineSpan> spans(String pinyin, {required TextStyle baseStyle}) {
    final spans = <InlineSpan>[];
    var i = 0;
    while (i < pinyin.length) {
      final c = pinyin[i];
      if (_isPinyinLetter(c)) {
        var j = i + 1;
        while (j < pinyin.length && _isPinyinLetter(pinyin[j])) {
          j++;
        }
        final run = pinyin.substring(i, j);
        for (final (syllable, tone) in _splitRun(run)) {
          spans.add(TextSpan(text: syllable, style: baseStyle.copyWith(color: colors[tone])));
        }
        i = j;
      } else {
        var j = i + 1;
        while (j < pinyin.length && !_isPinyinLetter(pinyin[j])) {
          j++;
        }
        spans.add(TextSpan(text: pinyin.substring(i, j), style: baseStyle));
        i = j;
      }
    }
    return spans;
  }
}

/// Convenience widget: renders [pinyin] with per-syllable tone coloring.
class TonedPinyinText extends StatelessWidget {
  final String pinyin;
  final TextStyle? style;
  final TextAlign? textAlign;

  const TonedPinyinText(this.pinyin, {super.key, this.style, this.textAlign});

  @override
  Widget build(BuildContext context) {
    final baseStyle = style ?? DefaultTextStyle.of(context).style;
    return RichText(
      textAlign: textAlign ?? TextAlign.start,
      text: TextSpan(children: PinyinTone.spans(pinyin, baseStyle: baseStyle)),
    );
  }
}

File: android/app/src/main/kotlin/com/osmankutlu/zh_en_dict/PinyinTone.kt
package com.osmankutlu.zh_en_dict

import android.os.Build
import android.text.Html
import android.text.Spanned

/**
 * Tone-colors Chinese text for the home-screen widget, mirroring the in-app
 * Dart `PinyinTone` palette and syllable segmentation so the widget matches
 * the flashcards. Word pinyin in the bundled data is written joined
 * (e.g. "tóngshì"), so we segment it into syllables via a standard
 * initials/finals table and read the tone off whichever vowel carries a
 * diacritic.
 *
 * Output is HTML (`<font color=...>`) fed through [fromHtml] into a
 * RemoteViews TextView. If text and pinyin can't be aligned we return null so
 * the caller falls back to plain (uncolored) text rather than risk wrong
 * colors.
 */
object PinyinTone {

    // Same hex values as lib/utils/pinyin_tone.dart.
    private val toneColor = mapOf(
        1 to "#1E88E5", // blue
        2 to "#43A047", // green
        3 to "#F9A825", // yellow
        4 to "#E53935", // red
        5 to "#9E9E9E", // grey - neutral
    )

    private val toneOfCharLower = mapOf(
        'ā' to 1, 'á' to 2, 'ǎ' to 3, 'à' to 4,
        'ē' to 1, 'é' to 2, 'ě' to 3, 'è' to 4,
        'ī' to 1, 'í' to 2, 'ǐ' to 3, 'ì' to 4,
        'ō' to 1, 'ó' to 2, 'ǒ' to 3, 'ò' to 4,
        'ū' to 1, 'ú' to 2, 'ǔ' to 3, 'ù' to 4,
        'ǖ' to 1, 'ǘ' to 2, 'ǚ' to 3, 'ǜ' to 4,
    )

    private val toneOfChar: Map<Char, Int> =
        toneOfCharLower + toneOfCharLower.mapKeys { it.key.uppercaseChar() }

    private val plainOfCharLower = mapOf(
        'ā' to 'a', 'á' to 'a', 'ǎ' to 'a', 'à' to 'a',
        'ē' to 'e', 'é' to 'e', 'ě' to 'e', 'è' to 'e',
        'ī' to 'i', 'í' to 'i', 'ǐ' to 'i', 'ì' to 'i',
        'ō' to 'o', 'ó' to 'o', 'ǒ' to 'o', 'ò' to 'o',
        'ū' to 'u', 'ú' to 'u', 'ǔ' to 'u', 'ù' to 'u',
        'ǖ' to 'v', 'ǘ' to 'v', 'ǚ' to 'v', 'ǜ' to 'v',
        'ü' to 'v', 'Ü' to 'v',
    )

    private val plainOfChar: Map<Char, Char> =
        plainOfCharLower + plainOfCharLower.mapKeys { it.key.uppercaseChar() }

    private val initials = listOf(
        "zh", "ch", "sh",
        "b", "p", "m", "f", "d", "t", "n", "l",
        "g", "k", "h", "j", "q", "x", "r", "z", "c", "s", "y", "w",
    ).sortedByDescending { it.length }

    private val finals = listOf(
        "iang", "iong", "uang", "ueng",
        "ian", "iao", "ing", "ong", "uai", "uan", "van",
        "ai", "ei", "ao", "ou", "an", "en", "ang", "eng", "er", "ue",
        "ia", "ie", "iu", "in", "ua", "uo", "ui", "un", "ve",
        "a", "o", "e", "i", "u", "v",
    ).sortedByDescending { it.length }

    private fun isPinyinLetter(c: Char): Boolean {
        val ascii = (c in 'a'..'z') || (c in 'A'..'Z')
        return ascii || toneOfChar.containsKey(c) || plainOfChar.containsKey(c)
    }

    private fun plain(c: Char): Char = plainOfChar[c] ?: c.lowercaseChar()

    private fun stripTones(s: String): String = s.map { plain(it) }.joinToString("")

    private fun toneOf(syllable: String): Int {
        for (c in syllable) toneOfChar[c]?.let { return it }
        return 5
    }

    /** Splits one run of pinyin letters into (syllableText, tone) pairs. */
    private fun splitRun(run: String): List<Pair<String, Int>> {
        val plain = run.map { plain(it) }.joinToString("")
        val result = mutableListOf<Pair<String, Int>>()
        var pos = 0
        while (pos < run.length) {
            var matchedInitial = ""
            for (init in initials) {
                if (plain.startsWith(init, pos)) { matchedInitial = init; break }
            }
            val afterInitial = pos + matchedInitial.length
            var matchedFinal = ""
            for (fin in finals) {
                if (plain.startsWith(fin, afterInitial)) { matchedFinal = fin; break }
            }
            if (matchedFinal.isEmpty()) {
                val chunk = run.substring(pos, pos + 1)
                // A lone trailing "r" is the 儿化 (erhua) tail of the previous
                // syllable, not a syllable of its own.
                if ((chunk == "r" || chunk == "R") && result.isNotEmpty()) {
                    val last = result.removeAt(result.size - 1)
                    result.add(Pair(last.first + chunk, last.second))
                } else {
                    result.add(Pair(chunk, toneOf(chunk)))
                }
                pos += 1
                continue
            }
            val end = afterInitial + matchedFinal.length
            val syllable = run.substring(pos, end)
            result.add(Pair(syllable, toneOf(syllable)))
            pos = end
        }
        return result
    }

    /** Flat, in-order list of every syllable as (text, tone) pairs. */
    private fun syllables(pinyin: String): List<Pair<String, Int>> {
        val result = mutableListOf<Pair<String, Int>>()
        var i = 0
        while (i < pinyin.length) {
            if (isPinyinLetter(pinyin[i])) {
                var j = i + 1
                while (j < pinyin.length && isPinyinLetter(pinyin[j])) j++
                result.addAll(splitRun(pinyin.substring(i, j)))
                i = j
            } else {
                i++
            }
        }
        return result
    }

    private fun isCjk(c: Char): Boolean {
        val code = c.code
        return (code in 0x4E00..0x9FFF) || (code in 0x3400..0x4DBF)
    }

    private fun esc(c: Char): String = when (c) {
        '&' -> "&amp;"
        '<' -> "&lt;"
        '>' -> "&gt;"
        else -> c.toString()
    }

    private fun span(color: String?, body: String): String =
        if (color == null) body else "<font color=\"$color\">$body</font>"

    /**
     * HTML for [hanzi] with each CJK character colored by the tone of its
     * aligned pinyin syllable. Returns null if there's no pinyin or the
     * syllable count doesn't line up with the characters.
     */
    fun hanziHtml(hanzi: String, pinyin: String?): String? {
        if (pinyin.isNullOrEmpty()) return null
        val syl = syllables(pinyin)
        if (syl.isEmpty()) return null
        val sb = StringBuilder()
        var s = 0
        var lastWasErhua = false
        var lastColor: String? = null
        for (c in hanzi) {
            if (!isCjk(c)) { sb.append(esc(c)); continue }
            if (c == '儿' && lastWasErhua) {
                sb.append(span(lastColor, esc(c)))
                lastWasErhua = false
                continue
            }
            if (s >= syl.size) return null // misaligned — don't risk wrong colors
            val (text, tone) = syl[s]
            val color = toneColor[tone]
            sb.append(span(color, esc(c)))
            lastColor = color
            val plainSyl = stripTones(text).lowercase()
            lastWasErhua = plainSyl.length > 1 && plainSyl.endsWith("r") && plainSyl != "er"
            s++
        }
        return sb.toString()
    }

    /** HTML for [pinyin] with each syllable colored by its tone. */
    fun pinyinHtml(pinyin: String?): String? {
        if (pinyin.isNullOrEmpty()) return null
        val sb = StringBuilder()
        var i = 0
        while (i < pinyin.length) {
            val c = pinyin[i]
            if (isPinyinLetter(c)) {
                var j = i + 1
                while (j < pinyin.length && isPinyinLetter(pinyin[j])) j++
                for ((syllable, tone) in splitRun(pinyin.substring(i, j))) {
                    sb.append(span(toneColor[tone], syllable.map { esc(it) }.joinToString("")))
                }
                i = j
            } else {
                sb.append(esc(c))
                i++
            }
        }
        return sb.toString()
    }

    @Suppress("DEPRECATION")
    fun fromHtml(html: String): Spanned =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
            Html.fromHtml(html, Html.FROM_HTML_MODE_LEGACY)
        else
            Html.fromHtml(html)
}
